test/validation/api/ipsec/ipsec_test_out.c

```diff
@@ -55,6 +55,38 @@ static struct auth_param auths[] = {
 	ALG(ODP_AUTH_ALG_AES_XCBC_MAC, &key_5a_128, NULL)
 };
 
+struct cipher_auth_comb_param {
+	struct cipher_param cipher;
+	struct auth_param auth;
+};
+
+static struct cipher_auth_comb_param cipher_auth_comb[] = {
+	{
+		ALG(ODP_CIPHER_ALG_AES_GCM, &key_a5_128, &key_mcgrew_gcm_salt_2),
+		ALG(ODP_AUTH_ALG_AES_GCM, NULL, NULL),
+	},
+	{
+		ALG(ODP_CIPHER_ALG_AES_GCM, &key_a5_192, &key_mcgrew_gcm_salt_2),
+		ALG(ODP_AUTH_ALG_AES_GCM, NULL, NULL),
+	},
+	{
+		ALG(ODP_CIPHER_ALG_AES_GCM, &key_a5_256, &key_mcgrew_gcm_salt_2),
+		ALG(ODP_AUTH_ALG_AES_GCM, NULL, NULL),
+	},
+	{
+		ALG(ODP_CIPHER_ALG_NULL, NULL, NULL),
+		ALG(ODP_AUTH_ALG_AES_GMAC, &key_a5_128, &key_mcgrew_gcm_salt_2),
+	},
+	{
+		ALG(ODP_CIPHER_ALG_NULL, NULL, NULL),
+		ALG(ODP_AUTH_ALG_AES_GMAC, &key_a5_192, &key_mcgrew_gcm_salt_2),
+	},
+	{
+		ALG(ODP_CIPHER_ALG_NULL, NULL, NULL),
+		ALG(ODP_AUTH_ALG_AES_GMAC, &key_a5_256, &key_mcgrew_gcm_salt_2),
+	},
+};
```

Comment:
This could cover all combined mode algorithms used in ESP (as the commit message suggests). AES-CCM and CHACHA20-POLY1305 are now missing. Adding those too and removing the individual tests will also make the later patches a tad smaller.

Reply:
Agreed. Only issue is, CCM support has some problem. Capabilities says CCM support is there but the current unit test itself skips the tests after printing "Wrong auth capabilities"

Since I'm covering all algos for all tests, the log would look like,

```
    ODP_CIPHER_ALG_NULL (keylen 0) ODP_AUTH_ALG_AES_GMAC (keylen 256) Wrong auth capabilities
Wrong auth capabilities
Wrong auth capabilities

    ODP_CIPHER_ALG_CHACHA20_POLY1305 (keylen 256) ODP_AUTH_ALG_CHACHA20_POLY1305 (keylen 0)
  passed
  Test: test_ah_out_in_all ...
    ODP_AUTH_ALG_NULL (keylen 0)
    ODP_AUTH_ALG_MD5_HMAC (keylen 128)
    ODP_AUTH_ALG_SHA1_HMAC (keylen 160)
    ODP_AUTH_ALG_SHA256_HMAC (keylen 256)
    ODP_AUTH_ALG_SHA384_HMAC (keylen 384)
    ODP_AUTH_ALG_SHA512_HMAC (keylen 512)
    ODP_AUTH_ALG_AES_XCBC_MAC (keylen 128)
  passed
  Test: test_ipsec_stats ...
        Stats : successWrong auth capabilities
Wrong auth capabilities
Wrong auth capabilities

        Stats : proto errWrong auth capabilities
Wrong auth capabilities
Wrong auth capabilities

  passed
```
I'll submit this in the next version.

Reply:
Yes, I know there is a problem with CCM in linux-gen IPsec. The short story is that it is not supported in linux-gen even though odp_ipsec_capability() claims so.

The long story:

The IPsec implementation in Linux-gen wants to use AES-CCM with 16 byte ICV length but the AES-CCM implementation in the crypto subsystem does not support that ICV length. The IPsec implementation does not realize this in the capability call and returns AES-CCM in the list of supported algorithms. Then odp_ipsec_auth_capability() checks the ICV length and does not return anything for AES-CCM, giving rise to the error you see.

An underlying problem here is that the ODP IPsec API does not support user-specified ICV lengths when an algorithm (or its application to IPsec) supports multiple lengths and leaves it unspecified which length is used, AES-GCM supports multiple lengths, but only one (16 bytes) is mandatory, so the current API is ok'ish with the user having to deduce or assume that ICV will be 16 bytes. or by adding a note about the ICV length in the API. But AES-CCM has two mandatory ICV lengths, so fixing ODP API to support only one of them would not be very good and the only real solution seems to be to add provisions for the user to specify the ICV length. I am planning to submit a PR for such an API change.

In the mean time, I think you could just ignore the error message as it does not cause validation test failure (it probably should) and it is not a problem in the validation test itself.

Reply:
Understood. Thanks for the explanation.